### DISCUSSION TITLE
What is av_libglesv2.dll that it is now generated on Avalonia v11?

### DISCUSSION THREAD

**Discussion by Shivansps** [AvaloniaUI/Avalonia#12597] [Q&A] [answered]
I remember this file was only being generated for arm64 windows.

Im not enterely sure what it does but removing it seems to have no effect, aditionally i have one app that it is critical to avoid hardware acceleration because it works on a enviroment were GPU drivers are not installed and the app will stop working once the driver is installed and starts working. Ive found that incluiding a precompiled mesa for windows opengl.dll along with the .exe allows this and it seems to work well. BUT with avalonia v11 i must also delete av_libglesv2.dll otherwise the app will use hardware acceleration even if i have the opengl.dll that has the software renderer incluided.

**Comment by maxkatz6** [comment]
It's an [Angle library](https://github.com/AvaloniaUI/angle), specifically a fork with some minor changes.
Basically a thin layer between OGL API to Dx11 API.
Without av_libglesv2.dll there is no proper hardware rendering in Avalonia, it will fallback to software rendering.

**Reply by Shivansps** [reply]
Im not 100% convinced it runs on software whiout it, im assuming then this is a OPENGL ES to DX11 wrapper then?
This was not present on 0.10.x, i belived Avalonia used native OpenGL.

So it is safe to remove it then? It will just run on software and not cause unexpected crashes? I need to run on pure software in at least one of my apps.

**Reply by maxkatz6** [reply, marked as answer]
> This was not present on 0.10.x

It was there.

>  i belived Avalonia used native OpenGL.

There is an option to use WSL, but it's not enabled by default.

> So it is safe to remove it then? 

Yeah, but I would disable it in the Win32PlatformOptions too. See https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32PlatformOptions.cs#L91-L94
Something like this in the BuildAvaloniaApp method:
```
.With(new Win32PlatformOptions
{
     RenderingMode = new[]
    {
        Win32RenderingMode.Software
    }
})
```

**Reply by maxkatz6** [reply]
Though, in 0.10 you might have missing av_libglesv2.dll, if you have built a "broken build" without more specific RID.
I.e. this file wasn't included with "win-x64" before, but only with "win7-x64" RID.
There is no difference now.

**Reply by Shivansps** [reply]
It was only on win-arm64 for me on 0,10... anyway. thanks for the reply.

**Comment by Shivansps** [comment]
> > This was not present on 0.10.x
> 
> It was there.
> 
> > i belived Avalonia used native OpenGL.
> 
> There is an option to use WSL, but it's not enabled by default.
> 
> > So it is safe to remove it then?
> 
> Yeah, but I would disable it in the Win32PlatformOptions too. See https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32PlatformOptions.cs#L91-L94 Something like this in the BuildAvaloniaApp method:
> 
> ```
> .With(new Win32PlatformOptions
> {
>      RenderingMode = new[]
>     {
>         Win32RenderingMode.Software
>     }
> })
> ```
Ok i wanted to ask a follow up question. This works perfectly ive been using it for a while, now. There is no way change the rendering mode during runtime after the application has already started?

Also this option is for windows only i would guess, what happens on Linux/Mac?

**Reply by maxkatz6** [reply]
> There is no way change the rendering mode during runtime after the application has already started

No.

> Also this option is for windows only i would guess, what happens on Linux/Mac?

System OpenGL is used (EaGL on mac, EGL on Linux).

**Reply by maxkatz6** [reply]
You also can force software renderer there in a similar way though.

**Reply by thevortexcloud** [reply]
You can also force software rendering at the driver level in various ways on Linux.

**Reply by Shivansps** [reply]
Got it i see the x11 platform options and the avalinianativeplatforms options for linux and mac.

One final question. There is any way to check from code if the application is currently running on hardware or software mode? That would be only to display that information.

**Reply by maxkatz6** [reply]
I don't remember any reliable way to do so. We might want to have some diagnostic/informational API.
